File: solutions/Linked List/141. Linked List Cycle.java
//https://leetcode.com/problems/linked-list-cycle


// First Approach - Floyd's Cycle Detection Algorithm (Tortoise and Hare)
// Time Complexity: O(n) - Space Complexity: O(1)
public class Solution {
    public boolean hasCycle(ListNode head) {
        if(head == null || head.next == null)
            return false;

        ListNode slow = head;
        ListNode fast = head.next;

        while(slow != fast) {
            slow = slow.next;

            if(fast.next != null && fast.next.next != null)
                fast = fast.next.next;
            else
                return false;
        }

        return true;
    }
}

// Second Approach - Optimized
// Time Complexity: O(n) - Space Complexity: O(1)
public class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;

            if (slow == fast)
                return true;
        }

        return false;
    }
}

File: solutions/Linked List/141. Linked List Cycle.py
# https://leetcode.com/problems/linked-list-cycle

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """

        """
        Time Complexity:
        Space Complexity:

        Idea: 2 pointers fast (+2) and slow (+1)
        - If they are the same before reaching the end (null)
        -> Cycle = true
        """

        slow = fast = head

        while fast and fast.next: 
            fast = fast.next.next
            slow = slow.next

            if fast == slow:
                return True
        
    
        return False